lib/puppet/functions/foreman/foreman.rb

```diff
@@ -0,0 +1,141 @@
+#        foreman_pass => 'my_api_foreman_pass'
+#
+# 'item' may be: environments, fact_values, hosts, hostgroups, puppetclasses, smart_proxies, subnets
+# 'search' is your actual search query.
+# 'per_page' specifies the maximum number of results you'd like to receive.
+#            This defaults to '20' which is consistent with what you'd get from
+#            Foreman if you didn't specify anything.
+# 'foreman_url' is your actual foreman server address
+# 'foreman_user' is the username of an account with API access
+# 'foreman_pass' is the password of an account with API access
+# 'filter_result' string or array with attribites to return
+#                 if a string is given foreman() returns an array only with given attributes
+#                 in case of an array is given foreman() returns an array of hashes selecting only
+#                 attributes given in array
+#                 in case of an given hash foreman() returns an array of hashes selecting only
+#                 attribute keys given in hash renamed to values of given keys. This can be used
+#                 to rename keys in result
+# 'timeout' is the Foreman request timeout in seconds as an integer.
+#           This defaults to five seconds.
+#
+# Then, use a variable to capture its output:
+# $f = ['hosts', 'hostgroup=Grid', '20', 'https://foreman', 'my_api_foreman_user', 'my_api_foreman_pass']
+# $hosts = foreman(*$f)
+#
+# Note: If you're using this in a template, you may be receiving an array of
+# hashes. So you might need to use two loops to get the values you need.
+#
+# Happy Foreman API-ing!
+
+require "yaml"
+require "net/http"
+require "net/https"
+require "uri"
+require "timeout"
+
+Puppet::Functions.create_function(:'foreman::foreman') do
+  dispatch :foreman do
+    required_param 'String', :item
+    required_param 'String', :search
+    optional_param 'Variant[Integer, String]', :per_page
+    optional_param 'String', :foreman_url
+    optional_param 'String', :foreman_user
+    optional_param 'String', :foreman_pass
+    optional_param 'Integer', :timeout
+    optional_param 'Variant[String, Array, Hash, Boolean]', :filter_result
+    optional_param 'Variant[Boolean, String]', :use_tfmproxy
+  end
+
+  def foreman(item, search, per_page = "20", foreman_url = "https://localhost", foreman_user = "admin", foreman_pass = "changeme", timeout = 5, filter_result = false, use_tfmproxy = false)
+    # extend this as required
+    searchable_items = %w{ environments fact_values hosts hostgroups puppetclasses smart_proxies subnets }
+    raise Puppet::ParseError, "Foreman: Invalid item to search on: #{item}, must be one of #{searchable_items.join(", ")}." unless searchable_items.include?(item)
+    raise Puppet::ParseError, "Foreman: Invalid filter_result: #{filter_result}, must be a String or an Array or false" unless filter_result.is_a? String or filter_result.is_a? Array or filter_result.is_a? Hash or filter_result == false
```

Comment:
Isn't this redundant now?

Reply:
It would fail if you pass `filter_result = true`. Puppet would accept it because its a boolean but we can't use it.

Reply:
Since it's optional already, would it make sense not to accept a boolean at all then or simplify the error here to only reject true?

Reply:
`false` is the default value for this parameters, so it should accept Booleans, but we can obviously simplify the error.